File: src/types/Culture.ts
import * as DB from "tsondb/schema/dsl"
import { CommonnessRatedAdvantageDisadvantage } from "./_CommonnessRatedAdvantageDisadvantage.js"
import {
  AdvantageIdentifier,
  BlessedTraditionIdentifier,
  DisadvantageIdentifier,
  LanguageIdentifier,
  MagicalTraditionIdentifier,
  ProfessionIdentifier,
  ProfessionVariantIdentifier,
  ScriptIdentifier,
  SkillIdentifier,
  SocialStatusIdentifier,
} from "./_Identifier.js"
import { BinarySex } from "./_Sex.js"
import { NestedTranslationMap } from "./Locale.js"
import { MundaneProfessionGroup } from "./Profession.js"
import { Errata } from "./source/_Erratum.js"
import { src } from "./source/_PublicationRef.js"

export const Culture = DB.Entity(import.meta.url, {
  name: "Culture",
  namePlural: "Cultures",
  type: () =>
    DB.Object({
      language: DB.Required({
        comment: "A list of native languages (usually it is only one).",
        type: DB.Array(LanguageIdentifier(), { minItems: 1, uniqueItems: true }),
      }),
      script: DB.Optional({
        comment:
          "A list of native scripts (usually it is only one). If the culture does not use any script, leave this array empty.",
        type: DB.Array(ScriptIdentifier(), { minItems: 1, uniqueItems: true }),
      }),
      area_knowledge: DB.Required({
        comment: "If the area knowledge has a fixed value or can be adjusted.",
        type: DB.IncludeIdentifier(AreaKnowledge),
      }),
      social_status: DB.Required({
        comment: "A list of possible social status in the respective culture.",
        type: DB.Array(SocialStatusIdentifier(), { minItems: 1, uniqueItems: true }),
      }),
      common_professions: DB.Required({
        comment:
          "A list of professions that are typical for the culture, as well as professions that are rarely practiced or encountered in the culture. The list is either defined by group (as multiple lists) or plain (as a single list).",
        type: DB.IncludeIdentifier(CommonProfessions),
      }),
      common_advantages: DB.Optional({
        comment: "A list of common advantages.",
        type: DB.Array(
          DB.GenIncludeIdentifier(CommonnessRatedAdvantageDisadvantage, [AdvantageIdentifier()]),
          { minItems: 1 },
        ),
      }),
      common_disadvantages: DB.Optional({
        comment: "A list of common disadvantages.",
        type: DB.Array(
          DB.GenIncludeIdentifier(CommonnessRatedAdvantageDisadvantage, [DisadvantageIdentifier()]),
          { minItems: 1 },
        ),
      }),
      uncommon_advantages: DB.Optional({
        comment: "A list of uncommon advantages.",
        type: DB.Array(
          DB.GenIncludeIdentifier(CommonnessRatedAdvantageDisadvantage, [AdvantageIdentifier()]),
          { minItems: 1 },
        ),
      }),
      uncommon_disadvantages: DB.Optional({
        comment: "A list of uncommon disadvantages.",
        type: DB.Array(
          DB.GenIncludeIdentifier(CommonnessRatedAdvantageDisadvantage, [DisadvantageIdentifier()]),
          { minItems: 1 },
        ),
      }),
      common_skills: DB.Required({
        comment: "A list of common skills.",
        type: DB.Array(DB.IncludeIdentifier(CommonnessRatedSkill), { minItems: 1 }),
      }),
      uncommon_skills: DB.Optional({
        comment: "A list of uncommon skills.",
        type: DB.Array(DB.IncludeIdentifier(CommonnessRatedSkill), { minItems: 1 }),
      }),
      cultural_package: DB.Required({
        comment: "The skill points you get for buying the culture package.",
        type: DB.Array(DB.IncludeIdentifier(CulturalPackageItem), { minItems: 1 }),
      }),
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "Culture",
        DB.Object({
          name: DB.Required({
            comment: "The race’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          area_knowledge: DB.Required({
            comment: "The description of the area knowledge.",
            type: DB.IncludeIdentifier(AreaKnowledgeTranslation),
          }),
          common_advantages: DB.Optional({
            comment: "The respective common advantages text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
          common_disadvantages: DB.Optional({
            comment: "The respective common disadvantages text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
          uncommon_advantages: DB.Optional({
            comment: "The respective uncommon advantages text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
          uncommon_disadvantages: DB.Optional({
            comment: "The respective uncommon disadvantages text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
          common_names: DB.Required({
            comment: "Structured description of common names.",
            type: DB.IncludeIdentifier(CommonNames),
          }),
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

const AreaKnowledge = DB.TypeAlias(import.meta.url, {
  name: "AreaKnowledge",
  comment: "If the area knowledge has a fixed value or can be adjusted.",
  type: () =>
    DB.Object({
      is_fixed: DB.Required({
        comment: "`true` if the area knowledge has a fixed value, `false` if it can be adjusted.",
        type: DB.Boolean(),
      }),
    }),
})

const CommonnessWeight = DB.Enum(import.meta.url, {
  name: "CommonnessWeight",
  comment:
    "The “weight” difference compared to other professions or profession variants. Some professions or profession variants are simply more common (`Mostly`), but sometimes only specific elements are used (`Only`).",
  values: () => ({
    Mostly: DB.EnumCase({ type: null }),
    Only: DB.EnumCase({ type: null }),
  }),
})

const Weighted = DB.GenTypeAlias(import.meta.url, {
  name: "Weighted",
  comment:
    "Some professions or profession variants are more common than others. There may be cultures where some professions or profession variants are not represented at all.",
  parameters: [DB.Param("ProfessionOrVariant")],
  type: ProfessionOrVariant =>
    DB.Object({
      elements: DB.Required({
        comment: "The list of more common professions or profession variants.",
        type: DB.Array(DB.TypeArgument(ProfessionOrVariant), { minItems: 1 }),
      }),
      weight: DB.Required({
        comment: "The list of more common professions or profession variants.",
        type: DB.IncludeIdentifier(CommonnessWeight),
      }),
    }),
})

const CommonProfessionConstraintsOperation = DB.Enum(import.meta.url, {
  name: "CommonProfessionConstraintsOperation",
  comment:
    "This defines how the list of constraints should be offset against the list of all mundane professions: Either only the professions are kept that intersect with the constraints or only the professions are kept that are different from the constraints.",
  values: () => ({
    Intersection: DB.EnumCase({ type: null }),
    Difference: DB.EnumCase({ type: null }),
  }),
})

const CommonProfessionConstraints = DB.GenTypeAlias(import.meta.url, {
  name: "CommonProfessionConstraints",
  comment:
    "A list of professions. The filter specifies how the list is applied to all mundane professions.",
  parameters: [DB.Param("Constraint")],
  type: Constraint =>
    DB.Object({
      constraints: DB.Required({
        comment: "The list of constraints.",
        type: DB.Array(DB.TypeArgument(Constraint), { minItems: 1 }),
      }),
      operation: DB.Required({
        comment:
          "This defines how the list of constraints should be offset against the list of all mundane professions: Either only the professions are kept that intersect with the constraints or only the professions are kept that are different from the constraints.",
        type: DB.IncludeIdentifier(CommonProfessionConstraintsOperation),
      }),
    }),
})

const Rarity = DB.Enum(import.meta.url, {
  name: "Rarity",
  comment: "Some professions may be found in a culture, but are not that common.",
  values: () => ({
    Rare: DB.EnumCase({ type: null }),
    VeryRare: DB.EnumCase({ type: null }),
  }),
})

const ProfessionConstraint = DB.TypeAlias(import.meta.url, {
  name: "ProfessionConstraint",
  type: () =>
    DB.Object({
      id: DB.Required({
        comment: "The profession’s identifier.",
        type: ProfessionIdentifier(),
      }),
      weighted_variants: DB.Optional({
        comment:
          "Some profession variants are more common than others. There may be cultures where some variants are not represented at all.",
        type: DB.GenIncludeIdentifier(Weighted, [ProfessionVariantIdentifier()]),
      }),
      rarity: DB.Optional({
        comment: "Some professions may be found in a culture, but are not that common.",
        type: DB.IncludeIdentifier(Rarity),
      }),
    }),
})

const MagicalTraditionConstraint = DB.TypeAlias(import.meta.url, {
  name: "MagicalTraditionConstraint",
  type: () =>
    DB.Object({
      id: DB.Required({
        comment: "The magical tradition’s identifier.",
        type: MagicalTraditionIdentifier(),
      }),
      weighted_professions: DB.Optional({
        comment:
          "Some professions are more common than others. There may be cultures where some professions are not represented at all.",
        type: DB.GenIncludeIdentifier(Weighted, [ProfessionIdentifier()]),
      }),
      rarity: DB.Optional({
        comment: "Some traditions may be found in a culture, but are not that common.",
        type: DB.IncludeIdentifier(Rarity),
      }),
    }),
})

const BlessedTraditionConstraint = DB.TypeAlias(import.meta.url, {
  name: "BlessedTraditionConstraint",
  type: () =>
    DB.Object({
      id: DB.Required({
        comment: "The blessed tradition’s identifier.",
        type: BlessedTraditionIdentifier(),
      }),
      weighted_professions: DB.Optional({
        comment:
          "Some professions are more common than others. There may be cultures where some professions are not represented at all.",
        type: DB.GenIncludeIdentifier(Weighted, [ProfessionIdentifier()]),
      }),
      rarity: DB.Optional({
        comment: "Some traditions may be found in a culture, but are not that common.",
        type: DB.IncludeIdentifier(Rarity),
      }),
    }),
})

const MundaneCommonProfessionConstraint = DB.Enum(import.meta.url, {
  name: "MundaneCommonProfessionConstraint",
  values: () => ({
    Profession: DB.EnumCase({ type: DB.IncludeIdentifier(ProfessionConstraint) }),
    ProfessionSubgroup: DB.EnumCase({ type: DB.IncludeIdentifier(MundaneProfessionGroup) }),
  }),
})

const MagicCommonProfessionConstraint = DB.Enum(import.meta.url, {
  name: "MagicCommonProfessionConstraint",
  values: () => ({
    Tradition: DB.EnumCase({ type: DB.IncludeIdentifier(MagicalTraditionConstraint) }),
    MagicDilettante: DB.EnumCase({ type: null }),
    Profession: DB.EnumCase({ type: DB.IncludeIdentifier(ProfessionConstraint) }),
  }),
})

const BlessedCommonProfessionConstraint = DB.Enum(import.meta.url, {
  name: "BlessedCommonProfessionConstraint",
  values: () => ({
    Tradition: DB.EnumCase({ type: DB.IncludeIdentifier(BlessedTraditionConstraint) }),
  }),
})

const PlainCommonProfessions = DB.TypeAlias(import.meta.url, {
  name: "PlainCommonProfessions",
  type: () => DB.GenIncludeIdentifier(CommonProfessionConstraints, [ProfessionIdentifier()]),
})

const GroupedCommonProfessions = DB.TypeAlias(import.meta.url, {
  name: "GroupedCommonProfessions",
  comment: "Lists of professions by group.",
  type: () =>
    DB.Object(
      {
        mundane: DB.Optional({
          type: DB.GenIncludeIdentifier(CommonProfessionConstraints, [
            DB.IncludeIdentifier(MundaneCommonProfessionConstraint),
          ]),
        }),
        magic: DB.Optional({
          type: DB.GenIncludeIdentifier(CommonProfessionConstraints, [
            DB.IncludeIdentifier(MagicCommonProfessionConstraint),
          ]),
        }),
        blessed: DB.Optional({
          type: DB.GenIncludeIdentifier(CommonProfessionConstraints, [
            DB.IncludeIdentifier(BlessedCommonProfessionConstraint),
          ]),
        }),
      },
      { minProperties: 1 },
    ),
})

const CommonProfessions = DB.Enum(import.meta.url, {
  name: "CommonProfessions",
  comment:
    "A list of professions that are typical for the culture, as well as professions that are rarely practiced or encountered in the culture. The list is either defined by group (as multiple lists) or plain (as a single list).",
  values: () => ({
    Plain: DB.EnumCase({ type: DB.IncludeIdentifier(PlainCommonProfessions) }),
    Grouped: DB.EnumCase({ type: DB.IncludeIdentifier(GroupedCommonProfessions) }),
  }),
})

const CommonnessRatedSkill = DB.TypeAlias(import.meta.url, {
  name: "CommonnessRatedSkill",
  type: () => SkillIdentifier(),
})

const CulturalPackageItem = DB.TypeAlias(import.meta.url, {
  name: "CulturalPackageItem",
  type: () =>
    DB.Object({
      id: DB.Required({
        comment: "The skill’s identifier.",
        type: SkillIdentifier(),
      }),
      points: DB.Required({
        comment:
          "The skill points for the respective skill you get for buying the cultural package.",
        type: DB.Integer({ minimum: 1, maximum: 2 }),
      }),
    }),
})

const AreaKnowledgeTranslation = DB.TypeAlias(import.meta.url, {
  name: "AreaKnowledgeTranslation",
  comment: "Description and examples of the area knowledge.",
  type: () =>
    DB.Object({
      description: DB.Required({
        comment: "The full description without examples in parenthesis.",
        type: DB.String({ minLength: 1 }),
      }),
      abbreviated: DB.Required({
        comment:
          "A shorter version of the description, used in input fields and other UI elements where the space might be to small to use the full description.",
        type: DB.String({ minLength: 1 }),
      }),
      examples: DB.Optional({
        comment: "Examples of areas, if applicable.",
        type: DB.Array(DB.IncludeIdentifier(AreaKnowledgeExample), { minItems: 1 }),
      }),
    }),
})

const AreaKnowledgeExample = DB.TypeAlias(import.meta.url, {
  name: "AreaKnowledgeExample",
  type: () =>
    DB.Object({
      area: DB.Required({
        type: DB.String({ minLength: 1 }),
      }),
    }),
})

const CommonNames = DB.TypeAlias(import.meta.url, {
  name: "CommonNames",
  comment: "Structured description of common names.",
  type: () =>
    DB.Object(
      {
        first_name_groups: DB.Optional({
          comment:
            "First names can be gender-neutral, but they can also be for a specific binary sex. They are sorted into groups.",
          type: DB.Array(DB.IncludeIdentifier(CommonNameGroup), { minItems: 1 }),
        }),
        last_name_groups: DB.Optional({
          comment:
            "Last names can be gender-neutral, like family names, but they can also be for a specific binary sex. They are sorted into groups.",
          type: DB.Array(DB.IncludeIdentifier(CommonNameGroup), { minItems: 1 }),
        }),
        naming_rules: DB.Optional({
          comment: "Special naming rules.",
          type: DB.String({ minLength: 1 }),
        }),
      },
      { minProperties: 1 },
    ),
})

const CommonNameGroup = DB.TypeAlias(import.meta.url, {
  name: "CommonNameGroup",
  type: () =>
    DB.Object(
      {
        label: DB.Required({
          comment: "The group label.",
          type: DB.String({ minLength: 1 }),
        }),
        sex: DB.Optional({
          comment: "The binary sex if the group is only for a certain binary sex.",
          type: DB.IncludeIdentifier(BinarySex),
        }),
        names: DB.Required({
          comment: "The names from the group.",
          type: DB.Array(DB.IncludeIdentifier(CommonName), { minItems: 1 }),
        }),
      },
      { minProperties: 1 },
    ),
})

const CommonName = DB.TypeAlias(import.meta.url, {
  name: "CommonName",
  type: () =>
    DB.Object(
      {
        name: DB.Required({
          type: DB.String({ minLength: 1 }),
        }),
        note: DB.Optional({
          comment: "Additional information about the name, appended in parenthesis.",
          type: DB.String({ minLength: 1 }),
        }),
      },
      { minProperties: 1 },
    ),
})

File: Sources/OptolithDatabaseSchema/Types/Culture.swift
import FileDB

@Model
public struct Culture {

  /// A list of native languages (usually it is only one).
  @MinItems(1)
  @UniqueItems
  @Relationship(Language.self)
  let language: [Language.ID]

  /// A list of native scripts (usually it is only one). If the culture does not use any script, leave this array empty.
  @MinItems(1)
  @UniqueItems
  @Relationship(Script.self)
  let script: [Script.ID]?

  /// If the area knowledge has a fixed value or can be adjusted.
  let area_knowledge: AreaKnowledge

  /// A list of possible social status in the respective culture.
  @MinItems(1)
  @UniqueItems
  @Relationship(SocialStatus.self)
  let social_status: [SocialStatus.ID]

  /// A list of professions that are typical for the culture, as well as professions that are rarely practiced or encountered in the culture. The list is either defined by group (as multiple lists) or plain (as a single list).
  let common_professions: CommonProfessions

      /// A list of common advantages.
      @MinItems(1)
      @Relationship(Advantage.self)
      let common_advantages: [CommonnessRatedAdvantageDisadvantage<Advantage.ID>]?

      /// A list of common disadvantages.
      @MinItems(1)
      @Relationship(Disadvantage.self)
      let common_disadvantages: [CommonnessRatedAdvantageDisadvantage<Disadvantage.ID>]?

      /// A list of uncommon advantages.
      @MinItems(1)
      @Relationship(Advantage.self)
      let uncommon_advantages: [CommonnessRatedAdvantageDisadvantage<Advantage.ID>]?

      /// A list of uncommon disadvantages.
      @MinItems(1)
      @Relationship(Disadvantage.self)
      let uncommon_disadvantages: [CommonnessRatedAdvantageDisadvantage<Disadvantage.ID>]?

  /// A list of common skills.
  @Relationship(Skill.self)
  @MinItems(1)
  let common_skills: [Skill.ID]

  /// A list of uncommon skills.
  @Relationship(Skill.self)
  @MinItems(1)
  let uncommon_skills: [Skill.ID]?

  /// The skill points you get for buying the culture package.
  @MinItems(1)
  let cultural_package: [CulturalPackageItem]

    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation { // CultureTranslation

        /// The race’s name.
        @MinLength(1)
        let name: String

        /// The description of the area knowledge.
        let area_knowledge: AreaKnowledgeTranslation

        /// The respective common advantages text from the source book.
        @MinLength(1)
        let common_advantages: String?

        /// The respective common disadvantages text from the source book.
        @MinLength(1)
        let common_disadvantages: String?

        /// The respective uncommon advantages text from the source book.
        @MinLength(1)
        let uncommon_advantages: String?

        /// The respective uncommon disadvantages text from the source book.
        @MinLength(1)
        let uncommon_disadvantages: String?

        /// Structured description of common names.
        let common_names: CommonNames

        /// A list of errata for the entry in the specific language.
        @MinItems(1)
        let errata: [Erratum]?
    }
}

/// If the area knowledge has a fixed value or can be adjusted.
@Embedded
public struct AreaKnowledge {

  /// `true` if the area knowledge has a fixed value, `false` if it can be adjusted.
  let is_fixed: Bool
  }

/// The “weight” difference compared to other professions or profession variants. Some professions or profession variants are simply more common (`Mostly`), but sometimes only specific elements are used (`Only`).
@ModelEnum
public enum CommonnessWeight {
    case Mostly
    case Only
}

/// Some professions or profession variants are more common than others. There may be cultures where some professions or profession variants are not represented at all.
@Embedded
public struct Weighted<ProfessionOrVariant> {
  /// The list of more common professions or profession variants.
  @MinItems(1)
  let elements: [ProfessionOrVariant]

  /// The list of more common professions or profession variants.
  let weight: CommonnessWeight
  }

/// This defines how the list of constraints should be offset against the list of all mundane professions: Either only the professions are kept that intersect with the constraints or only the professions are kept that are different from the constraints.
@ModelEnum
public enum CommonProfessionConstraintsOperation {
    case Intersection
    case Difference
}

/// A list of professions. The filter specifies how the list is applied to all mundane professions.
@Embedded
public struct CommonProfessionConstraints<Constraint> {
  /// The list of constraints.
  @MinItems(1)
  let constraints: [Constraint]

  /// This defines how the list of constraints should be offset against the list of all mundane professions: Either only the professions are kept that intersect with the constraints or only the professions are kept that are different from the constraints.
  let operation: CommonProfessionConstraintsOperation
  }

/// Some professions may be found in a culture, but are not that common.
@ModelEnum
public enum Rarity {
    case Rare
    case VeryRare
}

@Embedded
public struct ProfessionConstraint {

  /// The profession’s identifier.
  @Relationship(Profession.self)
  let id: Profession.ID

  /// Some profession variants are more common than others. There may be cultures where some variants are not represented at all.
  @Relationship(ProfessionVariant.self)
  let weighted_variants: Weighted<ProfessionVariant.ID>?

  /// Some professions may be found in a culture, but are not that common.
  let rarity: Rarity?
  }

@Embedded
public struct MagicalTraditionConstraint {

  /// The magical tradition’s identifier.
  @Relationship(MagicalTradition.self)
  let id: MagicalTradition.ID

  /// Some professions are more common than others. There may be cultures where some professions are not represented at all.
  @Relationship(Profession.self)
  let weighted_professions: Weighted<Profession.ID>?

  /// Some traditions may be found in a culture, but are not that common.
  let rarity: Rarity?
  }

@Embedded
public struct BlessedTraditionConstraint {

  /// The blessed tradition’s identifier.
  @Relationship(BlessedTradition.self)
  let id: BlessedTradition.ID

  /// Some professions are more common than others. There may be cultures where some professions are not represented at all.
  @Relationship(Profession.self)
  let weighted_professions: Weighted<Profession.ID>?

  /// Some traditions may be found in a culture, but are not that common.
  let rarity: Rarity?
  }

@ModelEnum
public enum MundaneCommonProfessionConstraint {
    case Profession(ProfessionConstraint)
    case ProfessionSubgroup(MundaneProfessionGroup)
}

@ModelEnum
public enum MagicCommonProfessionConstraint {
    case Tradition(MagicalTraditionConstraint)
    case MagicDilettante
    case Profession(ProfessionConstraint)
}

@ModelEnum
public enum BlessedCommonProfessionConstraint {
    case Tradition(BlessedTraditionConstraint)
}

@TypeAlias
public struct PlainCommonProfessions {
  @Relationship(Profession.self)
  let list: CommonProfessionConstraints<Profession.ID>
}

/// Lists of professions by group.
@Embedded
@MinProperties(1)
public struct GroupedCommonProfessions {
    let mundane: CommonProfessionConstraints<MundaneCommonProfessionConstraint>?

    let magic: CommonProfessionConstraints<MagicCommonProfessionConstraint>?

    let blessed: CommonProfessionConstraints<BlessedCommonProfessionConstraint>?
}

/// A list of professions that are typical for the culture, as well as professions that are rarely practiced or encountered in the culture. The list is either defined by group (as multiple lists) or plain (as a single list).
@ModelEnum
public enum CommonProfessions {
    case Plain(PlainCommonProfessions)
    case Grouped(GroupedCommonProfessions)
}

@Embedded
public struct CulturalPackageItem {

  /// The skill’s identifier.
  @Relationship(Skill.self)
  let id: Skill.ID

  /// The skill points for the respective skill you get for buying the cultural package.
  @Minimum(1)
  @Maximum(2)
  let points: Int
  }

/// Description and examples of the area knowledge.
@Embedded
public struct AreaKnowledgeTranslation {

  /// The full description without examples in parenthesis.
  @MinLength(1)
  let description: String

  /// A shorter version of the description, used in input fields and other UI elements where the space might be to small to use the full description.
  @MinLength(1)
  let abbreviated: String

  /// Examples of areas, if applicable.
  @MinItems(1)
  let examples: [AreaKnowledgeExample]?
  }

@Embedded
public struct AreaKnowledgeExample {

      @MinLength(1)
      let area: String
  }

/// Structured description of common names.
@Embedded
@MinProperties(1)
public struct CommonNames {
    /// First names can be gender-neutral, but they can also be for a specific binary sex. They are sorted into groups.
    @MinItems(1)
    let first_name_groups: [CommonNameGroup]?

    /// Last names can be gender-neutral, like family names, but they can also be for a specific binary sex. They are sorted into groups.
    @MinItems(1)
    let last_name_groups: [CommonNameGroup]?

    /// Special naming rules.
    @MinLength(1)
    let naming_rules: String?
}

@Embedded
@MinProperties(1)
public struct CommonNameGroup {
    /// The group label.
    @MinLength(1)
    let label: String

    /// The binary sex if the group is only for a certain binary sex.
    let sex: BinarySex?

    /// The names from the group.
    @MinItems(1)
    let names: [CommonName]
}

@Embedded
@MinProperties(1)
public struct CommonName {
    @MinLength(1)
    let name: String

    /// Additional information about the name, appended in parenthesis.
    @MinLength(1)
    let names: String?
}
